### PR TITLE
travis: test both Stack and non-stack (Cabal)

### LINKED ISSUE #124: travis: Document why we test the GHC versions that we do
For example #82 explains that we test 7.6 because some versions of Linux use that. An example I found was that [Ubuntu trusty uses 7.6](http://bazaar.launchpad.net/~ubuntu-branches/ubuntu/trusty/haskell-platform/trusty/view/head:/debian/control). It would be good if we would express the rationale as comments in the .travis.yml file so that we may potentially answer questions such as "when is it safe to remove our support for 7.6?"

To be clear, not that I'm advocating removing support of 7.6 anytime soon. In fact I may even advocate for adding 7.4, since I found that [Ubuntu Precise](http://bazaar.launchpad.net/~ubuntu-branches/ubuntu/precise/haskell-platform/precise/view/head:/debian/control) ships with that.

Edit: Indeed I still would have advocated for 7.4 but experiments in #131 to try to get 7.4 working were unsuccessful. (no `mkdtemp` in `System.Posix.Temp` in the version of `unix` we get with 7.4).

### LINKED ISSUE #153: travis: Travis-CI Roadmap
Right now, we have 4 *Issues* ( ~~#119~~ , #124 , #129 and ~~#130~~ ) and 2 "Pull Requests" ( ~~#120~~ , ~~#128~~ ) related to *Travis* pending. Except for 129, that seems to require substantial changes, all the others appear to be simple and will probably be merged soon.

However, before solving 129, it would be useful to at least have an idea about what we plan to achieve, so we can focus our work in a more productive way.

I think that, ideally, those things would be desirable in *Travis*:

- [x] Test with multiple versions of GHC.
- [ ] Isolated, reproducible test environments.
- [ ] Automatic handling of dependencies ( Depends on #151 )
- [ ] Minimal maintenance on software updates.

The only idea I had until now was to use a list of dependencies to generate *stack projects* on-the-fly to pass to *Travis* test, using cache to reduce subsequent build times. It works surprisingly well, but only if GHC version >= 7.8.2.

It would still be possible to treat GHC 7.6.x as a special case and use *cabal*. But that's is an ugly solution and adds a lot of complexity.

Alternatively, we can keep using a global GHC installation with *cabal*, which is simple and easy, but is also less reproducible and doesn't allow us to enforce correct declaration of exercises' dependencies.

I would really love some feedback about this from as many people as possible.

So... anyone?

### PR DESCRIPTION
**Pull request by petertseng**
The work in #162 is advantageous because we can check our dependencies
and have hermetic builds (do not break on upstream changes).
    
One question about whether to merge it was the necessity to drop GHC
7.6.
    
This commit introduces to Travis the ability to test with either Stack
or without (the old way of doing things). We test 7.6, 7.8, 7.10, 8.0
with cabal-install as we did before, and also Stack as done in #162.

Closes #162 (is based on it)
Closes #124 (explanatory comments provided)
Closes #153 (based on #162 which also closed #153)

### CODE REVIEW

**Comment by petertseng** [comment]
No longer WIP, let's take a look.

So we'll have the hermetic builds and dependency checking provided by Stack, but do not leave 7.6 behind.

One big question is whether the extra complexity is maintainable, which we should discuss.

It is possible that we can have 7.6 be the *only* version that uses cabal-install (Note that the current iteration of this PR tests 7.6, 7.8, 7.10, 8.0 with cabal-install, which may potentially be gratuitous.)

File: .travis.yml [hunk 1]
```diff
@@ -1,50 +1,88 @@
...
 script:
-  - 'export PATH=/opt/cabal/$CABALVER/bin:/opt/ghc/$GHCVER/bin:$PATH'
-  - 'runhaskell -Wall ./_test/check-exercises.hs'
-  - './bin/fetch-configlet'
-  - './bin/configlet .'
+  - |
+    configlet .
+    case "$BUILD" in
+      stack)
+        set -e
```
Comment (rbasso) [inline .travis.yml]: If `configlet` fails, why continue testing?

File: _test/bootstrap.sh [hunk 2]
```diff
@@ -2,12 +2,6 @@
 # This installs a the prerequisites for a given GHC/Cabal config
 set -x
 set -e
-if [ ! -z "$GHCVER" ]; then
```
Comment (petertseng) [inline _test/bootstrap.sh]: no longer needed as now handled by `before_install`. Can make it a separate commit if desired, though it will be a bit tricky as there will be one commit that has `PATH` being set in `script/cabal`, then one that moves it up to `before_install`

File: .travis.yml (hunk 1 above)
Comment (petertseng) [inline .travis.yml]: I take it this means we should move `set -e` above `configlet .` I agree.

File: .travis.yml [hunk 3]
```diff
@@ -1,50 +1,88 @@
...
+        ;;
+    esac
+
 install:
-  - true
+  - |
+    travis_retry fetch-configlet
+    if [ "$BUILD" = "stack" ]; then
+      travis_retry curl -L https://www.stackage.org/stack/linux-x86_64 -o pack.tgz
+      tar xzf pack.tgz --wildcards --strip-components=1 -C ${HOME}/bin '*/stack'
```
Comment (rbasso) [inline .travis.yml]: `tar` will even if  `fetch-configlet` and `curl -L` fail?

**Comment by petertseng** [comment]
updated to move `set -e` above `configlet`.

```diff
$ git diff e4b68e6 14f272b
diff --git a/.travis.yml b/.travis.yml
index c9fb04f..759ba58 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -64,10 +64,10 @@ install:
 
 script:
   - |
+    set -e
     configlet .
     case "$BUILD" in
       stack)
-        set -e
         for exercise in ${TRAVIS_BUILD_DIR}/exercises/* ; do
             pushd ${exercise}
             # `stack --word-dir` fails if not targeting a subdirectory, so we just
```

File: .travis.yml (hunk 3 above)
Comment (petertseng) [inline .travis.yml]: It might be best to fix that in #162 but I can do it here too. Trying to decide whether I should amend it into 9b73dee05da4bf7132c89dbeeed3ea7245737d3c (yours), have a separate one, or amend it into 14f272b0ead0ee61242b6a7662a8b078147b563f (mine), but perhaps I split hairs here

File: _test/bootstrap.sh (hunk 2 above)
Comment (rbasso) [inline _test/bootstrap.sh]: I will try to completely rewrite `bootstrap.sh` to use `dependencies` if I get the time. In #162 I removed it but, if we want to keep tests with cabal, we need to at least to centralize the data in `dependencies.txt`.

File: .travis.yml (hunk 3 above)
Comment (petertseng) [inline .travis.yml]: ok, just added a commit on

File: _test/bootstrap.sh (hunk 2 above)
Comment (petertseng) [inline _test/bootstrap.sh]: Seems like a good idea.

File: .travis.yml (hunk 3 above)
Comment (rbasso) [inline .travis.yml]: In #162 I used distinct lines of `install` so that Travis automatically fails everything if a line fails. When you moved the commands inside the `if` Travis cannot check intermediary steps.

Comment (rbasso) [inline .travis.yml]: You don't have to base this PR on #162.
The change are so significant that maybe it's better to just base it directly on master.

Comment (petertseng) [inline .travis.yml]: aha! a very subtle distinction that I did not notice! Thanks for the catch, then.

Comment (petertseng) [inline .travis.yml]: in that case I'll remove the intermediate commit

Comment (petertseng) [inline .travis.yml]: Maybe I should use `&&` rather than `set -e` here, hmm

**Comment by petertseng** [comment]
Moved some things to separate lines if possible.

```diff
$ git diff faf9502 247f0cf
diff --git a/.travis.yml b/.travis.yml
index 1b596f4..5b8a467 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -40,8 +40,8 @@ matrix:
   fast_finish: true                     # is released, so we list them here.
 
 before_install:
+  - export PATH="${TRAVIS_BUILD_DIR}/bin:$PATH"       # For {,fetch-}configlet.
   - |
-    export PATH="${TRAVIS_BUILD_DIR}/bin:$PATH"       # For {,fetch-}configlet.
     case "$BUILD" in
       stack)
         mkdir -p ${HOME}/bin                          # Create folder for stack.
@@ -55,18 +55,18 @@ before_install:
     esac
 
 install:
+  - travis_retry fetch-configlet
   - |
     set -e
-    travis_retry fetch-configlet
     if [ "$BUILD" = "stack" ]; then
       travis_retry curl -L https://www.stackage.org/stack/linux-x86_64 -o pack.tgz
       tar xzf pack.tgz --wildcards --strip-components=1 -C ${HOME}/bin '*/stack'
     fi
 
 script:
+  - configlet .
   - |
     set -e
-    configlet .
     case "$BUILD" in
       stack)
         for exercise in ${TRAVIS_BUILD_DIR}/exercises/* ; do
```

File: .travis.yml [hunk 4]
```diff
@@ -1,50 +1,89 @@
...
+
 script:
-  - 'export PATH=/opt/cabal/$CABALVER/bin:/opt/ghc/$GHCVER/bin:$PATH'
-  - 'runhaskell -Wall ./_test/check-exercises.hs'
-  - './bin/fetch-configlet'
-  - './bin/configlet .'
+  - configlet .
+  - |
+    set -e
+    case "$BUILD" in
```
Comment (rbasso) [inline .travis.yml]: In `script:` Travis has a special behavior. If a command fails, Travis will fail everything, but it will not shortcut evaluation, so every entry will be executed.

I used `set -e` and only one command in #162 to avoid that.

Comment (petertseng) [inline .travis.yml]: Ah interesting. Would we like this to happen? Consider this case:

If I'm adding a new exercise and have forgotten to add it to config.json AND its tests are failing: Maybe I want to see both of those things. If we stop the build immediately on `configlet` failing, then I have to fix that first, and only after I do that will I find out my tests are failing too. Is that OK?

Then again, the argument may be "well maybe I should have run the tests locally to see whether they were passing" which... is valid. Maybe the tests only fail on an old version of GHC though?

Comment (petertseng) [inline .travis.yml]: editing into previous comment:

I'm OK with either of these behaviors, but we should make the explicit decision.

File: .travis.yml (hunk 3 above)
Comment (petertseng) [inline .travis.yml]: OK, I'll not base it on #162 so that _bootstrap.sh doesn't get removed and then re-added, which was the main weird thing about basing this on #162.

File: .travis.yml (hunk 4 above)
Comment (rbasso) [inline .travis.yml]: I prefer to fail ASAP:

- Most of the time we are waiting for Travis, so faster is better.
- Usually the changes are small, so we have just one error.
- Travis is just a final check, it should be tested locally.
- It saves resources for other projects.
- It marginally decreases global warming. :grimacing:

**Comment by petertseng** [comment]
OK, I pretty much squashed #162 and this together, this avoids the delete and then re-add of bootstrap.sh. The only noticeable change is I also tried to add explanatory comments. The functionality as Travis sees it should not change.

File: .travis.yml (hunk 4 above)
Comment (petertseng) [inline .travis.yml]: OK, in that case I'll move configlet back into the single command, after `set -e`

**Comment by petertseng** [comment]
Almost certainly has merge conflicts with #179, feel free to merge #179 first

**Comment by petertseng** [comment]
Rebased now that #179 is in.

I'll assume we want to keep the `cabal-install` 7.8, 7.10, 8.0 unless I hear otherwise.

It's worth asking because, well, if the only reason we want to keep the `cabal-install` in Travis is to keep 7.6, then we should ask whether we are getting value out of the 7.8, etc.
It's possible there are other reasons we keep them though. Need to think that a bit more.

**Comment by rbasso** [comment]
About keeping the other GHC versions with cabal, I don't see the need, but that doesn't bother me at all.
We can leave the way it is just to see if we get luck and just one of the tests fail. That would be interesting...
